community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/sort_nodes.py

```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env python3
+
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This script sorts nodes based on their `physicalHost`.
+
+See https://cloud.google.com/compute/docs/instances/use-compact-placement-policies
+
+You can reduce latency in tightly coupled HPC workloads (including distributed ML training) 
+by deploying them to machines that are located close together. 
+For example, if you deploy your workload on a single physical rack, you can expect lower latency 
+than if your workload is spread across multiple racks. 
+Sending data across multiple rack requires sending data through additional network switches. 
+
+Example usage:
+``` my_sbatch.sh
+#SBATCH --ntasks-per-node=8
+#SBATCH --nodes=64
+
+export SLURM_HOSTFILE=$(/slurm/scripts/sort_nodes.py)
+
+srun -l hostname | sort
+```
+"""
+import os
+import subprocess
+import uuid
+from typing import List, Optional, Dict, Callable
+
+def order(paths: List[List[str]]) -> List[str]:
+    """
+    Orders the leafs of the tree in a way that minimizes the sum of distance in between 
+    each pair of neighboring nodes in the resulting order.
+    The resulting order will always start from the first node in the input list.
+    The ordering is "stable" with respect to the input order of the leafs i.e.
+    given a choice between two nodes (identical in other ways) it will select "nodelist-smallest" one.
+    """
+    if not paths: return []
+    # lookup table for node names order
+    name_idx = {p[-1]: i for i, p in enumerate(paths)}
+    class Vert:
+        "Represents a vertex in a *network* tree."
+        def __init__(self, name: str, parent: "Vert"):
+            self.name = name
+            self.parent = parent
+            self.children = {}
+            self.lex_key: int = -1
+
+    # build a tree
+    root = Vert("", None)
+    for path in paths:
+        n = root
+        for v in path:
+            if v not in n.children:
+                n.children[v] = Vert(v, n)
+            n = n.children[v]
+
+    # propagate "lexicographical key" from the leaves to the root
+    def set_lex_key(v: Vert) -> None:
+        if not v.children: # this is a node
+            v.lex_key = name_idx[v.name]
+        else:
+            for u in v.children.values():
+                set_lex_key(u)
+            v.lex_key = min(u.lex_key for u in v.children.values())
+    set_lex_key(root)
```

Comment:
I don't see where lex_key is ever used, which I think is what you want as dict maintains arg input order where desired.

Reply:
Good catch. Looking into it, lead to refactoring, see the diff.